### DISCUSSION TITLE
Minus button on the "add reminder" activity does not return to the original date

### DISCUSSION THREAD

**Discussion by dumblob** [felixwiemuth/SimpleReminder#37] [Ideas and feedback]
Please describe the expected and observed behaviour. You can also add screenshots.

Please add the following information:
- App version: 0.9.11 (latest on f-droid)
- Android version: 7.1

## Steps to reproduce

1. click "Add" icon in your launcher
2. press "+" button top right
3. press "-" button top left

## Expected behavior

The date number will return to the original (which is being shown before any interaction with the activity).

## Current behavior

The date number will **not** return to the original but stay with the "current +1" number.

**Comment by felixwiemuth** [comment]
Hi @dumblob,

It is actually intended to work as it does and how you describe it.

If you try, you will see that you can add a reminder for any desired time at the current day.

After trying a bit more and reading the help text in the app, does it make sense or do you have any suggestion how it could be described better?

It might not feel that intuitive that clicking "+" and then "-" does not result in the same UI state as before. If anything, one would change the initial state though, but the current initial state can be seen as more intuitive (showing current day) and also has an advantage regarding testing (see help). This is the compromise I chose.

**Comment by dumblob** [comment]
Thanks for the quick reply and a nudge.

I might still disappoint you somewhat :wink:.

Despite I somewhat understand the motivation, I am still certain it could (should) be implemented differently UX-wise (google "undo" - feel free to start with https://en.wikipedia.org/wiki/Mode_(user_interface) or maybe https://www.uxmatters.com/mt/archives/2020/03/are-you-sure-versus-undo-design-and-technology.php etc.). The current behavior is truly problematic as it makes it truly difficult to correct a mistake or "change your mind" which happens often during the state of mind which this app focuses on (namely "quickly capture an idea").

OTOH I am not a pro user of SimpleReminder (I am a total newcomer) so my suggestion is not representative. For me this issue crossed the line of "trust" and I am reluctant to use it in my everyday life as I would be too scared to not touch certain parts of the display (and be afraid there are other similar or even worse "easter eggs" in the app elsewhere).

Feel free to close this issue if you do not plan to give it a second chance :wink:.

**Comment by felixwiemuth** [comment]
Hi again, thanks for some further explanation of your point.

To first address your worry regarding correcting mistakes: You *do* correct your choice by clicking the "-" button after having clicked the "+" button first. It *only* makes a difference in the initial state and if you do not change the time at all before adding the reminder: without clicking "+" and then "-" (and thus not having given any date/time input), it will be a test reminder which will be shown directly, and is thus a rather uncommon use case. With "+" and then "-", it is a reminder with the same time for the next day. In other words: if you click the "+" button and then change your mind and click the "-" button, you are fine and basically in the same state as before. If you also choose a time for the reminder, which you certainly want, there is no difference between the two scenarios. So you won't accidentally choose a wrong date for the reminder by correcting a "+/-" button choice. Details to understand why follow below.

I had thought quite a bit about the behaviour regarding date and time input, and I still like the current compromise best. But I am always open to better suggestions. I'll move this to the discussions.

You can read a bit about the evolution of this choice in #10.
But let me explain the key points here.

First of all, there are two date selection modes: "Next 24" (the selected time implicitly refers to its next occurrence within the next 24 hours) and "manual" (the date is explicitly selected via +/- buttons or calendar).
- It can always be read off from the UI in which mode we are:
  - Date in normal text color (and a blue "+1" if applicable): Next 24 mode
  - Date in orange: manual mode
- Initially, the UI is in Next 24 mode (but, as an exception for the initial state, it is assumed that the current time displayed refers to the current day - therefore there is no "+1").
  - I think this is preferable, as it would look strange to start with the next day (with a "+1"). Especially regarding that setting a reminder for the same day is probably quite common, and in that case one would prefer that the current day's date would be shown from the beginning, instead of the next day with a "+1" which first vanishes when one changes the time.
  - As a side-effect, this also allows for conveniently testing reminders: by not changing the time, the current time is chosen, which, when adding the reminder, is in the past and the reminder will be shown directly.
- From the initial state, the "+/-" buttons behave as follows:
  - Clicking the "+" button switches to manual mode (indicated by orange).
  - Clicking the "-" button afterwards switches back to Next 24 mode. Here it depends on which time is currently selected whether a "+1" will be shown or not. If one didn't change the time, it will be before the current day's time and thus the next day is selected ("+1" is shown). Thus a slight difference to the initial state.
- It might be worth noting that in general the dynamic component of time plays a role here: whether the selected time is within the next 24 hours or not (and which date this corresponds to) can change while the UI is not manipulated, and it will update on the next change made to time or date.
  - With this, you can get the same asymmetric behaviour from a not-initial state: Assume it is 12:00, set time to 12:01 (current date will be displayed), wait until it is 12:01 (still current date is displayed - if you add the reminder now, it will be due in the past and shown directly, same as in initial state), click the "+" button (next day and "+1" will be displayed), click the "-" button, "+1" will still be displayed.
  - I think this behaviour is definitely preferable over the UI (and the selected date) automatically updating while time passes but no input is made.
  - When this behaviour exists anyway, it can even be considered consistent that the initial state also behaves like this. One could say that the initial UI state corresponds to an initialization before the minute turned to the current, but still with that minute chosen (because at that point in time that minute was in the future, and thus the current day should be referred to).

In summary I would say that because of the many possible cases it is a somewhat complex, but still consistent behaviour.

There exists actually one other, deliberate type of asymmetry: when in manual mode with "+1" where the selected time is before the current day's time, the behaviour is as follows:
- Starting from Next 24 mode with "+1", clicking the "+" button leads to manual mode with "+2". This is chosen because from a "+1" you want to move to "+2" directly (the alternative would be that it switches to "+1", manual mode). The current way better supports the use case where you first chose a time in the initial Next 24 mode, with the date being "+1", and then change to the day after ("+2").
- However, clicking the "-" button then leads to "+1" still in manual mode, instead of back in Next 24 mode. The reasoning here is that once one has entered manual mode (by clicking on the "+/-" buttons or using the calendar), it is assumed that one wants to choose the date manually, and then it would be bad to land in the Next 24 mode again, where the date can change automatically when changing the time.
- Only a second click of the "-" button leads back to Next 24 mode; this is simply to allow to navigate back to Next 24 mode at all.
- To make it symmetric, the "-" button would have to lead from manual mode "+2" to either manual mode "+1" (if selected time is before the current day's time) or Next 24 mode "+1" otherwise. This is maybe not so intuitive on first view either. But especially following the reasoning that having once used the date controls one should set the date manually, I prefer the current way.

Some general design goals for the app that play a role here:
- Select date and time efficiently (quickly, with minimum amount of clicks, ...) - therefore no separate selection of date and time which of course would be simplest.
- Optimize for long-term, not for first time use: rather spend some time getting to know the app, then wasting time each time using it.
- Optimize for "forward navigation" (the shortest, most direct or most intuitive ways of setting a date and time) and not for "backward navigation" (correcting a choice).

In general I think that it's always good to try every feature a few times and get to know them before relying on them.

All in all, I think that the current behaviour is good, but it could definitely need some more explanation to get familiar with it. Eventually I think that an interactive intro, explaining how the mode switching behaves, would be a good thing. I am prioritizing other important features for now, but if anyone wants to do this, a contribution would be welcome.

**Reply by dumblob** [reply]
Oh, thanks for the precise and deep explanation and the pointer to #10 . I appreciate your effort and think it actually solves the goals you have outlined quite well (and for now I could not come up with any better solution).

But a highly visual intro, either a video or animation or even interactive "learn by accomplishing small tasks one by one" would solve the issue.

Feel free to "close" this discussion as I feel everything is clear to me now. Will look forward to the intro/video/animation/tasks enhancement.

Thanks again for pioneering such app!
